File: lua/test/ticket_type_entity_test.lua
-- TicketType entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("intercom_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("TicketTypeEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:TicketType(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["ticket_type"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:TicketType(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:TicketType(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = ticket_type_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "ticket_type." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set INTERCOM_TEST_TICKET_TYPE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local ticket_type_ref01_ent = client:TicketType(nil)
    local ticket_type_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.ticket_type"), "ticket_type_ref01"))

    local ticket_type_ref01_data_result, err = ticket_type_ref01_ent:create(ticket_type_ref01_data, nil)
    assert.is_nil(err)
    ticket_type_ref01_data = helpers.to_map(type(ticket_type_ref01_data_result) == 'table' and ticket_type_ref01_data_result.data_get and ticket_type_ref01_data_result:data_get() or ticket_type_ref01_data_result)
    assert.is_not_nil(ticket_type_ref01_data)
    assert.is_not_nil(ticket_type_ref01_data["id"])

    -- LIST
    local ticket_type_ref01_match = {}

    local ticket_type_ref01_list_result, err = ticket_type_ref01_ent:list(ticket_type_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(ticket_type_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(ticket_type_ref01_list_result),
      { id = ticket_type_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local ticket_type_ref01_data_up0_up = {
      id = ticket_type_ref01_data["id"],
    }

    local ticket_type_ref01_markdef_up0_name = "category"
    local ticket_type_ref01_markdef_up0_value = "Mark01-ticket_type_ref01_" .. tostring(setup.now)
    ticket_type_ref01_data_up0_up[ticket_type_ref01_markdef_up0_name] = ticket_type_ref01_markdef_up0_value

    local ticket_type_ref01_resdata_up0_result, err = ticket_type_ref01_ent:update(ticket_type_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local ticket_type_ref01_resdata_up0 = helpers.to_map(type(ticket_type_ref01_resdata_up0_result) == 'table' and ticket_type_ref01_resdata_up0_result.data_get and ticket_type_ref01_resdata_up0_result:data_get() or ticket_type_ref01_resdata_up0_result)
    assert.is_not_nil(ticket_type_ref01_resdata_up0)
    assert.are.equal(ticket_type_ref01_resdata_up0["id"], ticket_type_ref01_data_up0_up["id"])
    assert.are.equal(ticket_type_ref01_resdata_up0[ticket_type_ref01_markdef_up0_name], ticket_type_ref01_markdef_up0_value)

    -- LOAD
    local ticket_type_ref01_match_dt0 = {
      id = ticket_type_ref01_data["id"],
    }
    local ticket_type_ref01_data_dt0_loaded, err = ticket_type_ref01_ent:load(ticket_type_ref01_match_dt0, nil)
    assert.is_nil(err)
    local ticket_type_ref01_data_dt0_load_result = helpers.to_map(type(ticket_type_ref01_data_dt0_loaded) == 'table' and ticket_type_ref01_data_dt0_loaded.data_get and ticket_type_ref01_data_dt0_loaded:data_get() or ticket_type_ref01_data_dt0_loaded)
    assert.is_not_nil(ticket_type_ref01_data_dt0_load_result)
    assert.are.equal(ticket_type_ref01_data_dt0_load_result["id"], ticket_type_ref01_data["id"])

  end)
end)

function ticket_type_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/ticket_type/TicketTypeTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read ticket_type test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "ticket_type01", "ticket_type02", "ticket_type03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("INTERCOM_TEST_TICKET_TYPE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["INTERCOM_TEST_TICKET_TYPE_ENTID"] = idmap,
    ["INTERCOM_TEST_LIVE"] = "FALSE",
    ["INTERCOM_TEST_EXPLAIN"] = "FALSE",
    ["INTERCOM_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["INTERCOM_TEST_TICKET_TYPE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["INTERCOM_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["INTERCOM_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["INTERCOM_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["INTERCOM_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/ticket_type_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/intercom-sdk/go"
	"github.com/voxgig-sdk/intercom-sdk/go/core"

	vs "github.com/voxgig-sdk/intercom-sdk/go/utility/struct"
)

func TestTicketTypeEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.TicketType(nil)
		if ent == nil {
			t.Fatal("expected non-nil TicketTypeEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"ticket_type": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.TicketType(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.TicketType(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := ticket_typeBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "ticket_type." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set INTERCOM_TEST_TICKET_TYPE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		ticketTypeRef01Ent := client.TicketType(nil)
		ticketTypeRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "ticket_type"}), "ticket_type_ref01"))

		ticketTypeRef01DataResult, err := ticketTypeRef01Ent.Create(ticketTypeRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		ticketTypeRef01Data = core.ToMapAny(entityData(ticketTypeRef01DataResult))
		if ticketTypeRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if ticketTypeRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		ticketTypeRef01Match := map[string]any{}

		ticketTypeRef01ListResult, err := ticketTypeRef01Ent.List(ticketTypeRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		ticketTypeRef01List, ticketTypeRef01ListOk := ticketTypeRef01ListResult.([]any)
		if !ticketTypeRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", ticketTypeRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(ticketTypeRef01List), map[string]any{"id": ticketTypeRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		ticketTypeRef01DataUp0Up := map[string]any{
			"id": ticketTypeRef01Data["id"],
		}

		ticketTypeRef01MarkdefUp0Name := "category"
		ticketTypeRef01MarkdefUp0Value := fmt.Sprintf("Mark01-ticket_type_ref01_%d", setup.now)
		ticketTypeRef01DataUp0Up[ticketTypeRef01MarkdefUp0Name] = ticketTypeRef01MarkdefUp0Value

		ticketTypeRef01ResdataUp0Result, err := ticketTypeRef01Ent.Update(ticketTypeRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		ticketTypeRef01ResdataUp0 := core.ToMapAny(entityData(ticketTypeRef01ResdataUp0Result))
		if ticketTypeRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if ticketTypeRef01ResdataUp0["id"] != ticketTypeRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if ticketTypeRef01ResdataUp0[ticketTypeRef01MarkdefUp0Name] != ticketTypeRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", ticketTypeRef01MarkdefUp0Name, ticketTypeRef01ResdataUp0[ticketTypeRef01MarkdefUp0Name])
		}

		// LOAD
		ticketTypeRef01MatchDt0 := map[string]any{
			"id": ticketTypeRef01Data["id"],
		}
		ticketTypeRef01DataDt0Loaded, err := ticketTypeRef01Ent.Load(ticketTypeRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		ticketTypeRef01DataDt0LoadResult := core.ToMapAny(entityData(ticketTypeRef01DataDt0Loaded))
		if ticketTypeRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if ticketTypeRef01DataDt0LoadResult["id"] != ticketTypeRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func ticket_typeBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "ticket_type", "TicketTypeTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read ticket_type test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse ticket_type test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"ticket_type01", "ticket_type02", "ticket_type03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("INTERCOM_TEST_TICKET_TYPE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"INTERCOM_TEST_TICKET_TYPE_ENTID": idmap,
		"INTERCOM_TEST_LIVE":      "FALSE",
		"INTERCOM_TEST_EXPLAIN":   "FALSE",
		"INTERCOM_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["INTERCOM_TEST_TICKET_TYPE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["INTERCOM_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["INTERCOM_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewIntercomSDK(core.ToMapAny(mergedOpts))
	}

	live := env["INTERCOM_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["INTERCOM_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
